kv/kvnemesis: TestKVNemesisSingleNode failed

kv/kvnemesis.TestKVNemesisSingleNode [failed](https://tanzanite.cluster.engflow.com/invocations/default/8a61169b-7e6e-400e-b4ba-8cf5402af27d?testReportRun=4&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZuZW1lc2lzOmt2bmVtZXNpc190ZXN0) on release-24.1 @ [0d7110a85b5bbf0eb68bafd15abca076948b434e](https://github.com/cockroachdb/cockroach/commits/0d7110a85b5bbf0eb68bafd15abca076948b434e):


```
=== RUN   TestKVNemesisSingleNode
    test_log_scope.go:170: test logs captured to: outputs.zip/logTestKVNemesisSingleNode1704786730
    test_log_scope.go:81: use -show-logs to present logs inline
    kvnemesis_test.go:329: seed: 8784179012823337516
    kvnemesis_test.go:236: kvnemesis logging to /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_4_of_25/test.outputs/kvnemesis2045786910
    kvnemesis.go:178: committed read committed txn non-atomic timestamps: [sp(create)]id:0 [d]/Table/100/"09d904a032462975":missing-><nil>@s24 [r]/Table/100/"2ddb9c617445824b":[1718175979.786431858,0, 1718175979.811471110,1)->v32 [d]/Table/100/"97dcf51473d8b438":missing-><nil>@s25 [sp(rollback)]id:0 [r]/Table/100/"ae1aab78fa9cbb7a":[1718175979.822198818,0, <max>)->v40
    kvnemesis.go:213: failures(verbose): /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_4_of_25/test.outputs/kvnemesis2045786910/failures
        repro steps: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_4_of_25/test.outputs/kvnemesis2045786910/repro.go
        rangefeed KVs: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_4_of_25/test.outputs/kvnemesis2045786910/kvs-rangefeed.txt
        scan KVs: /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_4_of_25/test.outputs/kvnemesis2045786910/kvs-scan.txt
    kvnemesis_test.go:357: [0] proposed: 481
    kvnemesis_test.go:359: [0] reproposed unchanged: 64
    kvnemesis_test.go:361: [0] reproposed with new LAI: 2
    kvnemesis_test.go:366: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:366
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/kvnemesis_test.go:276
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisSingleNode
        	Messages:   	kvnemesis detected failures
    panic.go:626: -- test log scope end --
test logs left over in: outputs.zip/logTestKVNemesisSingleNode1704786730
--- FAIL: TestKVNemesisSingleNode (9.13s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=4</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisSingleNode.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-39503

--- a/pkg/kv/kvnemesis/validator.go
+++ b/pkg/kv/kvnemesis/validator.go
@@ -1115,8 +1115,17 @@ func (v *validator) checkAtomicCommitted(
 					panic(err)
 				}
 			}
+		case *observedRead:
+			// If v.observationFilter == observeLocking, this must be a locking read
+			// from a weak-isolation transaction. If this locking read is rolled back
+			// by a savepoint, there is no guarantee that the lock will be present
+			// after the savepoint rollback. Since the read is not part of a
+			// serializatble transaction and is no longer locking, it's valid at all
+			// times. Alternatively, we can remove the read from txnObservations.
+			if rollbackSp != nil && v.observationFilter == observeLocking {
+				o.ValidTimes = disjointTimeSpans{{Start: hlc.MinTimestamp, End: hlc.MaxTimestamp}}
+			}
 		case *observedSavepoint:
-
 			switch o.Type {
 			case create:
 				// Set rollbackSp to nil if this savepoint create matches the rolled
@@ -1188,7 +1197,11 @@ func (v *validator) checkAtomicCommitted(
 				}
 			}
 		case *observedRead:
-			o.ValidTimes = validReadTimes(batch, o.Key, o.Value.RawBytes, o.SkipLocked /* missingKeyValid */)
+			// For rolled back locking reads from weak-isolation transactions, the
+			// ValidTimes are already set above.
+			if len(o.ValidTimes) == 0 {
+				o.ValidTimes = validReadTimes(batch, o.Key, o.Value.RawBytes, o.SkipLocked /* missingKeyValid */)
+			}
 		case *observedScan:
 			// All kvs should be within scan boundary.
 			for _, kv := range o.KVs {

